crates/circuit/src/equivalence.rs

```diff
@@ -0,0 +1,844 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use itertools::Itertools;
+use smallvec::{smallvec, SmallVec};
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::{error::Error, fmt::Display};
+
+use exceptions::CircuitError;
+use hashbrown::{HashMap, HashSet};
+use pyo3::types::PyDict;
+use pyo3::{prelude::*, types::IntoPyDict};
+
+use rustworkx_core::petgraph::{
+    graph::{DiGraph, EdgeIndex, NodeIndex},
+    visit::EdgeRef,
+};
+
+use crate::circuit_instruction::{
+    convert_py_to_operation_type, CircuitInstruction as CircuitInstructionBase,
+    OperationTypeConstruct,
+};
+use crate::imports::ImportOnceCell;
+use crate::operations::Param;
+use crate::operations::{Operation, OperationType};
+
+mod exceptions {
+    use pyo3::import_exception_bound;
+    import_exception_bound! {qiskit.circuit.exceptions, CircuitError}
+}
+pub static PYDIGRAPH: ImportOnceCell = ImportOnceCell::new("rustworkx", "PyDiGraph");
+
+// Custom Structs
+
+#[pyclass(sequence, module = "qiskit._accelerate.circuit.equivalence")]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Key {
+    #[pyo3(get)]
+    pub name: String,
+    #[pyo3(get)]
+    pub num_qubits: u32,
+}
+
+#[pymethods]
+impl Key {
+    #[new]
+    #[pyo3(signature = (name="".to_string(), num_qubits=0))]
+    fn new(name: String, num_qubits: u32) -> Self {
+        Self { name, num_qubits }
+    }
+
+    fn __eq__(&self, other: Self) -> bool {
+        self.eq(&other)
+    }
+
+    fn __hash__(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        (self.name.to_string(), self.num_qubits).hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn __repr__(slf: PyRef<'_, Self>) -> String {
+        slf.to_string()
+    }
+
+    fn __getstate__(slf: PyRef<Self>) -> (String, u32) {
+        (slf.name.to_owned(), slf.num_qubits)
```

Comment:
I might do this
```suggestion
        (slf.name.clone(), slf.num_qubits)
```
because, unlike with `to_owned`, I know the output type immediately